### ISSUE TITLE
"bake" flag documentation / caching control for loaders

### ISSUE THREAD

**Issue by zzzeek** [sqlalchemy/sqlalchemy#6072]
the "enable_baked_queries" flag on Session has misleading documentation.  it only applies to the legacy "Baked" query extension which is seldom used and is no longer used by the ORM.   so this should be clarified.

next, the "bake_queries" flag on relationship() only has partial effects with the lazy loader; it will cancel the internal "lambda" process but does not actually disable query caching.   the docs for this flag should be chagned to reflect that this disables a partiular internal thing, but to turn off caching, this should be done engine-wide.   or if bake_queries is False then we send compiled_cache=None:

```diff
diff --git a/lib/sqlalchemy/orm/strategies.py b/lib/sqlalchemy/orm/strategies.py
index 51f75baf3..6566fa620 100644
--- a/lib/sqlalchemy/orm/strategies.py
+++ b/lib/sqlalchemy/orm/strategies.py
@@ -970,7 +970,10 @@ class LazyLoader(AbstractRelationshipLoader, util.MemoizedSlots):
                 stmt,
                 primary_key_identity,
                 load_options=load_options,
-                execution_options={"compiled_cache": self._query_cache},
+                execution_options={"compiled_cache": self._query_cache
+                        if self.parent_property.bake_queries else None
+
+                },
             )
 
         if self._order_by:

```

finally, there realy should be a way to disable caching for real, in the engine.  right now you can disable ORM query caching like this:

```
from sqlalchemy import event
@event.listens_for(Session, "do_orm_execute")
def do_orm_execute(execute_state):
    execute_state.update_execution_options(compiled_cache=None)


```

but an engine that has query_cache_size=0 should probably refuse to do any caching for anyone, including lazy loaders that send another cache in.   or we have a whole new flag "enable_caching=False", which means any compiled_caches passed in are also ignored.   we would need to note this at query_cache_size as well.      it's good to be able to fully turn the caching off entirely for debugging.

**Comment by zzzeek** [comment]
note also that LazyLoader is not using the strategy-local statement cache for loads that are not use_get, apparently.   We should likely move to having LazyLoader/SelectInLoader use the regular engine cache for statement caching and their own local cache just for the lambda part of it.

**Comment by gj** [comment]
> right now you can disable ORM query caching like this:
> 
> ```
> from sqlalchemy import event
> @event.listens_for(Session, "do_orm_execute")
> def do_orm_execute(execute_state):
>     execute_state.update_execution_options(compiled_cache=None)
> ```

I'm trying to disable caching on 1.4.15, and I can't get the above to work. The `compiled_cache` option doesn't seem to be updated to `None`, and this breakpoint is hit:

```py
@event.listens_for(Session, "do_orm_execute")
def do_orm_execute(execute_state):
    execute_state.update_execution_options(compiled_cache=None)
    after = execute_state.execution_options.get("compiled_cache")
    if after is not None:
        breakpoint()
```

**Comment by zzzeek** [comment]
> ```python
> @event.listens_for(Session, "do_orm_execute")
> def do_orm_execute(execute_state):
>     execute_state.update_execution_options(compiled_cache=None)
>     after = execute_state.execution_options.get("compiled_cache")
>     if after is not None:
>         breakpoint()
> ```

cant reproduce, please run the test case below and illustrate how to make it fail:

```python
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


@event.listens_for(Session, "do_orm_execute")
def do_orm_execute(execute_state):
    execute_state.update_execution_options(compiled_cache=None)
    after = execute_state.execution_options.get("compiled_cache")
    if after is not None:
        breakpoint()


Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)

    bs = relationship("B")


class B(Base):
    __tablename__ = 'b'
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))
    data = Column(String)


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A(bs=[B()]))

s.commit()

for a in s.query(A).all():
    a.bs
```


also, what is the reason you need to disable caching?   caching needs to be 100% bug free, and if not, you need to post an issue here describing what you are seeing.

**Comment by gj** [comment]
> cant reproduce, please run the test case below, modify it to produce an [MCVE](http://stackoverflow.com/help/mcve) illustrating how to make it fail:

Can't reproduce via modifying your snippet either — I'm probably doing something dumb in the actual code where I'm seeing the issue. Too tired to continue debugging now, but I'll update this thread tomorrow once I figure it out.

> also, what is the reason you need to disable caching?

Building authorization software (https://github.com/osohq/oso/tree/main/languages/python/sqlalchemy-oso) where query caching can lead to authorization bypass. On SQLAlchemy 1.3, I was able to get around this by setting `enable_baked_queries=False`, but I'm currently trying to update the library so that it supports SQLAlchemy 1.4 in addition to 1.3. This is one of a few assertions in the test suite that fails if baked queries are enabled on 1.3:

https://github.com/osohq/oso/blob/1f8cbcf49f247451e4f8e427bab57f10c83ad8c0/languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py#L331

Basically, a lazy-loaded relationship is queried via a session that has no authorization applied:

https://github.com/osohq/oso/blob/1f8cbcf49f247451e4f8e427bab57f10c83ad8c0/languages/python/sqlalchemy-oso/tests/test_sqlalchemy.py#L317-L318

And then when a subsequent query is made via the `AuthorizedSession`, which subclasses `Session` and sets `enable_baked_queries=False`, the relationship query returns `None` since authorization is correctly enforced.

**Comment by zzzeek** [comment]
1.3's baked queries had problems when events were used to modify them on the fly.  the new query caching with do_orm_execute() should solve this problem entirely.     you can modify the query in do_orm_execute() in any way that you'd like, and the cache key is only created **after** the query has been modified, so whatever modifications have been made will also be reflected in the cache key.   

I really need you to show me a self-contained example of how you are altering the query in such a way that caching is providing a stale query.

**Comment by gj** [comment]
Turns out — to no one's surprise — that it was late and I was being dumb. I missed this caveat in the `QueryEvents.before_compile` deprecation notice:

https://github.com/sqlalchemy/sqlalchemy/blob/71a858817cb8c11451ae577c61329f4239fab46b/lib/sqlalchemy/orm/events.py#L2676-L2679

Previously, we were using the `QueryEvents.before_compile` hook to apply authorization both on top-level queries as well as relationship loads, but the event is no longer emitted for attribute loads in 1.4.

So the main failure I was seeing was caused by not applying authorization to attribute loads instead of incorrect caching. I _was_ still seeing some cases where there was a populated cache even after setting the `compiled_cache=None` execution option, but I'm assuming that's another PEBKAC. If I see that behavior again, I will try to modify your test case to reproduce it, but I'm assuming it's all related to my incomplete update from 1.3 -> 1.4.

**Comment by zzzeek** [comment]
OK I hope to see oso doing well with the new architecture, oso's use case is basically exactly what do_orm_execute() had in mind.

**Comment by zzzeek** [comment]
re: caching control / bake flag, linked to #6487 the plan I'm thinking is:

1. adjust LRU caches in loader strategies to be 30 for lazy loader, 15 for selectinloader
2. loader strategy LRU is only used for lambda caches, no longer has a warning
3. bake_queries only refers to the lambda cache, that's it
4. all statement caching moves to the engine wide cache and is controlled like all other statement caching.    Confirm that lazy loading is already doing this for non-use-get queries anyway.

**Comment by zzzeek** [comment]
additionally, selectinload is not using the local query cache for statement caching either, so this is mostly just the lambdas being put into this cache.

**Comment by zzzeek** [comment]
one thing we are not doing yet is propagating a per-statement query_cache for an ORM statement into the "post loaders".  subqueryload was doing that, but selectinload was not, I'm taking that out of subqueryload for consistency.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**simplify relationship caching options** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2866
